crates/iota-data-ingestion-core/src/executor.rs

```diff
@@ -66,41 +82,102 @@ impl<P: ProgressStore> IndexerExecutor<P> {
-                        reader_checkpoint_number = seq_number;
+                Some(worker_pool_progress_msg) = self.pool_progress_receiver.recv() => {
+                    match worker_pool_progress_msg {
+                        WorkerPoolMsg::Progress((task_name, sequence_number)) => {
+                            self.progress_store.save(task_name.clone(), sequence_number).await?;
+                            let seq_number = self.progress_store.min_watermark()?;
+                            if seq_number > reader_checkpoint_number {
+                                gc_sender.send(seq_number).await?;
+                                reader_checkpoint_number = seq_number;
+                            }
+                            self.metrics.data_ingestion_checkpoint.with_label_values(&[&task_name]).set(sequence_number as i64);
+                        }
+                        // Manages the graceful shutdown sequence of the entire indexer system.
+                        //
+                        // The shutdown process follows these steps:
+                        // 1. Token cancellation triggers:
+                        //    a. Individual workers in each pool:
+                        //       - Complete current checkpoint processing
+                        //       - Send final progress updates
+                        //       - Signal completion to their pool
+                        //    b. Worker pools:
+                        //       - Stop accepting new checkpoints
+                        //       - Process remaining progress messages
+                        //       - Wait for all their workers to finish
+                        //       - Send ShutDown message to executor
+                        //
+                        // 2. Executor main loop:
+                        //    - Continues processing Progress messages from pools
+                        //    - Tracks pool shutdowns via ShutDown messages
+                        //    - Once all pools report shutdown:
+                        //      a. Awaits all worker pool join handles
+                        //      b. Signals checkpoint reader to stop
+                        //      c. Awaits checkpoint reader completion
+                        //      d. Exits main loop
+                        //
+                        // This ensures hierarchical shutdown order:
+                        // 1. Workers (in parallel within each pool)
+                        // 2. Worker pools (in parallel)
+                        // 3. Checkpoint reader
+                        // 4. Executor main loop
+                        //
+                        // Guarantees:
+                        // - No work is interrupted mid-processing
+                        // - All progress is saved to storage
+                        // - All messages are processed in order
+                        // - All resources are properly cleaned up
```

Comment:
This is far to much information for clarifying a nested branch of code. As it also extends beyond the scope of the executor shutdown logic I would suggest adding into the library-level documentation (`lib.rs`) in a dedicated section. 

You can then update the `run` method documentation by adding a link to the library-level documentation for more info.

Reply:
Since the `README.md` has a section on graceful shutdown could that be enough?

Reply:
[fixed](https://github.com/iotaledger/iota/pull/4648/commits/af78d4b08780698c81f8610bb0683cee2d17ec59)